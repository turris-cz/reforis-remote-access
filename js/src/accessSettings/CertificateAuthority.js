/*
 * Copyright (C) 2019-2021 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";
import PropTypes from "prop-types";

import {
    Button,
    useAlert,
    useAPIDelete,
    API_STATE,
    formFieldsSize,
} from "foris";

import API_URLs from "API";

export const CA_STATUS = {
    MISSING: "missing",
    READY: "ready",
    GENERATING: "generating",
};

CertificateAuthority.propTypes = {
    authority: PropTypes.object.isRequired,
    accessEnabled: PropTypes.bool.isRequired,
    onSuccess: PropTypes.func.isRequired,
};

export default function CertificateAuthority({
    authority,
    accessEnabled,
    onSuccess,
}) {
    const [setAlert] = useAlert();

    const [deleteResponse, deleteCA] = useAPIDelete(API_URLs.authority);
    useEffect(() => {
        if (deleteResponse.state === API_STATE.SUCCESS) {
            onSuccess();
        } else if (deleteResponse.state === API_STATE.ERROR) {
            setAlert(_("Cannot delete certificate authority."));
        }
    }, [deleteResponse, onSuccess, setAlert]);

    if (authority.status !== CA_STATUS.READY) {
        return null;
    }

    return (
        <div className={formFieldsSize}>
            <h2>{_("Certificate Authority")}</h2>
            {accessEnabled ? (
                <p>
                    {_(
                        "You can't delete certificate authority while remote access is enabled. To delete it, you need to disable the access first."
                    )}
                </p>
            ) : (
                <>
                    <p>
                        {_(
                            "Your certificate authority is set up properly. Please note that if you delete it, all clients will have their access revoked."
                        )}
                    </p>
                    <div className="text-right">
                        <Button
                            onClick={() => deleteCA()}
                            loading={deleteResponse.state === API_STATE.SENDING}
                            forisFormSize
                        >
                            {_("Delete Certificate Authority")}
                        </Button>
                    </div>
                </>
            )}
        </div>
    );
}
