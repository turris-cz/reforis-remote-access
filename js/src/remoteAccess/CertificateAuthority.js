/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";
import PropTypes from "prop-types";

import {
    Button, useAlert, useAPIDelete, API_STATE,
} from "foris";

import API_URLs from "API";

export const CA_STATUS = {
    MISSING: "missing",
    READY: "ready",
    GENERATING: "generating",
};

CertificateAuthority.propTypes = {
    authority: PropTypes.object.isRequired,
    accessEnabled: PropTypes.bool.isRequired,
    onSuccess: PropTypes.func.isRequired,
};

export default function CertificateAuthority({ authority, accessEnabled, onSuccess }) {
    const [setAlert] = useAlert();

    const [deleteResponse, deleteCA] = useAPIDelete(API_URLs.authority);
    useEffect(() => {
        if (deleteResponse.state === API_STATE.SUCCESS) {
            onSuccess();
        } else if (deleteResponse.state === API_STATE.ERROR) {
            setAlert(_("Cannot delete certificate authority"));
        }
    }, [deleteResponse, onSuccess, setAlert]);

    if (authority.status !== CA_STATUS.READY) {
        return null;
    }

    return (
        <>
            <h3>{_("Certificate Authority")}</h3>
            {accessEnabled
                ? <p>{_("You can't delete certificate authority while remote access is enabled. In order to delete it you need to disable the access first.")}</p>
                : (
                    <>
                        <p>{_("Your certificate authority is set up properly. Please note that if you delete it all clients will have their access revoked.")}</p>
                        <Button
                            onClick={() => deleteCA()}
                            loading={deleteResponse.state === API_STATE.SENDING}
                            forisFormSize
                        >
                            {_("Delete certificate authority")}
                        </Button>
                    </>
                )}
        </>
    );
}
