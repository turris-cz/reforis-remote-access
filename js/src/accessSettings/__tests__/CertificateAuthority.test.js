/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import mockAxios from "jest-mock-axios";
import { WebSockets } from "foris";
import {
    render,
    getByText,
    wait,
    fireEvent,
} from "foris/testUtils/customTestRender";
import { mockSetAlert } from "foris/testUtils/alertContextMock";
import { mockJSONError } from "foris/testUtils/network";

import CertificateAuthority, { CA_STATUS } from "../CertificateAuthority";

describe("<CertificateAuthority />", () => {
    const webSockets = new WebSockets();
    const onSucces = jest.fn();
    let container;

    function renderComponent(authority = {}, accessEnabled = false) {
        return render(
            <CertificateAuthority
                ws={webSockets}
                authority={authority}
                accessEnabled={accessEnabled}
                onSuccess={onSucces}
            />
        );
    }

    function deleteCA() {
        fireEvent.click(getByText(container, "Delete Certificate Authority"));
    }

    describe("when remote access is disabled", () => {
        it("should render component", () => {
            ({ container } = renderComponent(
                { status: CA_STATUS.READY },
                true
            ));
            expect(container).toMatchSnapshot();
        });
    });

    describe("when remote access is enabled", () => {
        beforeEach(() => {
            ({ container } = renderComponent({ status: CA_STATUS.READY }));
        });

        it("should render component", () => {
            expect(container).toMatchSnapshot();
        });

        it("should send request when button is clicked", () => {
            deleteCA();
            expect(mockAxios.delete).toBeCalledWith(
                "/reforis/remote-access/api/authority",
                expect.anything()
            );
        });

        it("should handle error", async () => {
            deleteCA();
            mockJSONError();
            await wait(() => {
                expect(mockSetAlert).toHaveBeenCalledWith(
                    "Cannot delete certificate authority."
                );
            });
        });

        it("should handle success", async () => {
            deleteCA();
            mockAxios.mockResponse({});
            await wait(() => {
                expect(onSucces).toHaveBeenCalledTimes(1);
            });
        });
    });
});
