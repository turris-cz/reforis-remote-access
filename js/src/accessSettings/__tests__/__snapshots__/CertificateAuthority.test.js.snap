// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CertificateAuthority /> when remote access is disabled should render component 1`] = `
<div>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
  >
    <h2>
      Certificate Authority
    </h2>
    <p>
      You can't delete certificate authority while remote access is enabled. To delete it, you need to disable the access first.
    </p>
  </div>
</div>
`;

exports[`<CertificateAuthority /> when remote access is enabled should render component 1`] = `
<div>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
  >
    <h2>
      Certificate Authority
    </h2>
    <p>
      Your certificate authority is set up properly. Please note that if you delete it, all clients will have their access revoked.
    </p>
    <div
      class="text-right"
    >
      <button
        class="btn btn-primary col-sm-12 col-md-3 col-lg-2 d-inline-flex justify-content-center align-items-center"
        type="button"
      >
        <span>
          Delete Certificate Authority
        </span>
      </button>
    </div>
  </div>
</div>
`;
